Authenticator updates

hamchapman opened the pull request:
### What?

Make `authenticate` return an `ErrorResponse` if there was an error when generating a new token pair.

### Why?

Make it match how we want the server SDKs to behave.

----

- [x] README updated if you changed the API?

----

CC @pusher/sigsdk

pl commented on src/authenticator.ts:
```diff
@@ -63,7 +70,7 @@ export default class Authenticator {
   }
 
-  private authenticateWithRefreshToken(oldRefreshToken: string, options: AuthenticateOptions): AuthenticationResponse {
+  private authenticateWithRefreshToken(oldRefreshToken: string, options: AuthenticateOptions): AuthenticationResponse | ErrorResponse {
```
I think the change to using `ErrorResponse` instead of specialised exceptions is good, but I'm not sure about using the return type for that. I don't like exceptions too much either, to be honest. Maybe we should unify the success and error cases, if we expect users to just forward them to the client. WDYT?

hamchapman replied:
Potentially, but unify them how? Isn't specifying the return type as a union type, unification?

hamchapman commented:
@pl a few things for you to consider:

- I unified the auth responses into an `AuthenticationResponse` object, which now looks like:

```js
export class AuthenticationResponse {
  readonly status: number;
  readonly headers: Headers;
  readonly body: TokenResponse | ErrorResponse;

  constructor(options: {
    status: number,
    headers?: Headers,
    body: TokenResponse | ErrorResponse,
  }) {
    this.status = options.status;
    this.headers = options.headers || {};
    this.body = options.body;
  }
}
```

where `TokenResponse` and `ErrorResponse` look like:

```js
export interface TokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
  refresh_token: string;
}

export class ErrorResponse {
  readonly status: number;
  readonly headers: Headers;
  readonly error: string;
  readonly error_description: string;
  readonly error_uri?: string;

  constructor(options: {
    status: number,
    headers?: Headers,
    error: string,
    error_description: string,
    error_uri?: string,
  }) {
    this.status = options.status;
    this.headers = options.headers || {};
    this.error = options.error;
    this.error_description = options.error_description;
    this.error_uri = options.error_uri;
  }
}
```

- Does having the status in the `ErrorResponse` make sense, when it's already in the `AuthenticationResponse`? This is where I feel like it's perhaps the wrong error type to be using here, whereas it _does_ make sense to have `ErrorResponse` look like this when you've received a response from a request to a Pusher Platform property. 

Perhaps there could be a basic `ErrorBody` (akin to what is described in ESG) which could look like: 

```js
export interface ErrorBody {
  readonly error: string;
  readonly error_description: string;
  readonly error_uri?: string;
}
```

And then `ErrorResponse` could either extend this or wrap it somehow? 

WDYT?

- Previously the expected response payload in the case of a successful auth request was defined as:

```js
{
  access_token: {
    token: 'adsasd',
    expires_id: 1000
  },
  token_type: 'bearer';
  expires_in: 20000;
  refresh_token: 'cvbccvbb'
}
```

Although this was only sometimes respected and more often the payload looked like:

```js
{
  access_token: 'adsasd',
  token_type: 'bearer';
  expires_in: 20000;
  refresh_token: 'cvbccvbb'
}
```

Clients (specifically Chatkit ones, at the very least) do not currently expect the first form (nested `token` inside of `access_token`). 

Is there a specific expected format as specified by a spec (OAuth2? - I will check this to see if I can find out)?

- Should the refresh tokens be generated with an expiry? See https://github.com/pusher/pusher-platform-node/blob/8811a2d876f65fb40304e6c323a1f05bcd05893e/src/authenticator.ts#L118-L132

I feel like they should, given this: https://github.com/pusher/pusher-platform-node/blob/8811a2d876f65fb40304e6c323a1f05bcd05893e/src/authenticator.ts#L70-L77

pl commented:
1. As you suggested, I'd split out `ErrorBody` and use it in the `AuthenticationResponse` class.
2. I don't know where the first format came from. I believe OAuth2 uses the second form and that's what I'd recommend.
3. I'm not sure about expiry on refresh tokens. I think it's a good practice, but could be annoying for users. I'm even wondering whether you need them in Chatkit.

hamchapman commented:
Updated again. List of changes:

* Made invalid grant type have a status of 422, not 401.
* `authenticate` doesn't return a refresh token
* Added `authenticateWithRefreshToken` that returns a refresh token as part of the token payload
* Created `ErrorBody` to make `AuthenticationResponse`s which are about errors not contain a full `ErrorResponse` (and therefore duplicated status and headers fields)

pl commented:
Code looks good. The only thing I'd do is prefixing error types. Instead of `invalid_grant_type` we should use something like `token_provider/invalid_grant_type`.

hamchapman commented:
Nice - will do that. Thanks!
